### ISSUE TITLE
Firebase should be part of a dependency bundle by default

### ISSUE THREAD

**Issue by ColtonIdle** [autonomousapps/dependency-analysis-android-gradle-plugin#305]
I used this tool on my android project for the first time and it says that this dependency is unused. If I remove this dependency then my code doesn't compile. Not sure why this dependency is being flagged as safe for removal.

**Comment by ColtonIdle** [comment]
I will try to come up with a minimal repro but it might take some time.

**Comment by ColtonIdle** [comment]
Okay.

So it was simpler than I thought it was going to be!

File > New > New Project > Basic Activity

Add `implementation "com.google.firebase:firebase-analytics:17.6.0"`

Then in your activity do

```
val analytics: FirebaseAnalytics? = null
analytics!!.logEvent("test", null)
```

Add the dep analysis plugin and it will say

```
Unused dependencies which should be removed:
- implementation("com.google.firebase:firebase-analytics:17.6.0")
```

I comment out `implementation("com.google.firebase:firebase-analytics:17.6.0")`, run `./gradlew buildHealth` and it fails because of 

```
Unresolved reference: firebase
Unresolved reference: FirebaseAnalytics
```

So yeah. That's the issue.
TLDR. Add the dep. Add the two lines for FirebaseAnalytics and it'll repro.

**Comment by plastiv** [comment]
I don't believe it's a false positive per se. `com.google.firebase.analytics.FirebaseAnalytics` class is located at `com.google.android.gms:play-services-measurement-api:` artifact and `com.google.firebase:firebase-analytics:` looks like marker parent artifact without any implementation inside.

So google asks to add `firebase-analytics` at the docs and leaves open possibility for changing internal artifacts structure. While dependency-analysis plugin tells you that nothing from that artifact is used.

**Comment by ColtonIdle** [comment]
Please excuse my ignorance and lack of knowledge of gradle and dependencies here... but it sounds like what you're saying is that firebase-analytics is empty and it just defers it's actual implementation to com.google.android.gms:play-services-measurement-api?

Which is why the list of "Transitively used dependencies that should be declared directly as indicated:" includes

`- implementation("com.google.android.gms:play-services-measurement-api:17.6.0")`?

So my solution in this case is to just swap firebase-analytics with play-services-measurement?

If so, I think it would be cool somehow indicate that a "removal" of a dependency, should really be a "replace" of a dependency. How else would I know that these two are connected?

**Comment by autonomousapps** [comment]
This is a common problem that I've only partially solved. There's a concept that I call "logical dependencies", which may be a single "physical" artifact (like a single jar or aar), or may be a composite of multiple dependencies that are closely related. This case is an example of the latter. Another common one is kotlin-stdlib, which has -jdkX variants; it's common to declare only the -jdkX variant even though you may only use the core stdlib.

The problem is that it's impossible (maybe?) to know if a set of dependencies constitutes a logical dependency a priori. It may just be that the publication was misconfigured and is exposing its own implementation dependencies on the consumer's compile classpath, and if so, I strongly believe that you should directly declare that transitive dependency if you also use it directly.

I have begun whitelisting known logical dependencies (referred to as "dependency bundles" in the plugin's DSL), with kotlin-stdlib being the sole example at the moment. I may consider adding more. In the meantime, users may add their own on an ad hoc basis. You can find out how in the [wiki](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/wiki/Customizing-plugin-behavior#logical-dependency-groups-aka-dependency-bundles); if you need help doing that, just let me know.

**Comment by ColtonIdle** [comment]
The explanation is enough for me @autonomousapps Thank you! Learned something new!
